src/entities/entity.ts

```diff
@@ -0,0 +1,73 @@
+import { Component } from '../components/component';
+
+/*
+ * Custom error class for duplicate component errors.
+ */
+class DuplicateComponentError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'DuplicateComponentError';
+  }
+}
+
+/**
+ * Represents an entity in the system.
+ */
+export class Entity {
+  /**
+   * The unique identifier for the entity.
+   */
+  id: string;
+
+  /**
+   * A record of components associated with the entity.
+   */
+  components: Record<string, Component> = {};
+
+  /**
+   * Creates a new entity with the given identifier.
+   * @param {string} id - The unique identifier for the entity.
+   */
+  constructor(id: string) {
+    this.id = id;
+  }
+
+  /**
+   * Adds a component to the entity.
+   * @param {Component} component - The component to be added.
+   * @throws {DuplicateComponentError} If a component of the same type already exists.
+   */
+  addComponent(component: Component): void {
+    const componentName = component.constructor.name;
+    if (this.components[componentName]) {
+      this.handleDuplicateComponentError(component);
+    }
+    this.components[component.constructor.name] = component;
+  }
+
+  /**
+   * Handles the error when attempting to add a duplicate component.
+   * @param {Component} component - The component that caused the error.
+   * @throws {DuplicateComponentError} Always thrown to indicate the error.
+   * @private
+   */
+  private handleDuplicateComponentError(component: Component): void {
+    const componentName = component.constructor.name;
+    console.error(`Can't add duplicate component: ${componentName}`);
+    throw new DuplicateComponentError(
+      `Can't add duplicate component: ${componentName}`
+    );
+  }
+
+  /**
+   * Gets a component of the specified type.
+   * @template T - The type of the component.
+   * @param {new (...args: any[]) => T} componentConstructor - The constructor of the component type.
+   * @returns {T | undefined} The component of the specified type, or undefined if not found.
+   */
+  getComponent<T>(
```

Comment:
Feels wrong to be able to request any arbitrary Component for any Entity. The types of Components possessed by an Entity should be enforced by the Entity hierarchy. For example:

```
   Entity
      |
      |
Character Entity
      |           \
PlayerEntity     Enemy Entity
```

Where CharacterEntity has a PhysicalComponent, RenderComponent, and Status Component, and a PlayerEntity, in addition to those, has a WeaponComponent.

You'll need to narrow down the exact Entity types you pass through your update code, but I think it's worth it.

Reply:
I agree with the idea of introducing an entity hierarchy, especially as more entity types are introduced. However, its impact on filtering might be limited. In ECS architecture, systems typically prioritize component existence over entity types. Entities are viewed more as containers for components, emphasizing composition rather than inheritance.

In ECS, systems commonly check for the presence of specific components for efficiency. While an entity hierarchy can bring organizational benefits, its direct influence on system filtering may be less significant. The approach depends on the project's specific needs—balancing the simplicity of generic component checks with the structured organization that an entity hierarchy provides.

Reply:
I believe there is merit in mandating entities to possess specific components through inheritance, particularly during the initialization phase, as it enhances readability. This approach ensures clarity in the setup process. Simultaneously, I advocate for systems to operate on a generic Entity, affording them the flexibility to determine actions based on the components attached. This bifurcated strategy strikes a balance between clarity in entity structure and adaptability in system functionality.

Reply:
It's your call. But if you're going to keep some hierarchy, then I'd suggest pulling up the common elements of PlayerEntity and EnemyEntity into a shared parent class ("CharacterEntity"?).

Reply:
It's your call. But if you're going to keep some hierarchy, then I'd suggest pulling up the common elements of PlayerEntity and EnemyEntity into a shared parent class ("CharacterEntity"?).

Reply:
Done 413c23552d04c7f290c4d2d2a31846bb6460a692